src/main/java/jbash/environment/JBashEnvironment.java

```diff
@@ -0,0 +1,121 @@
+    private final HashMap<String, String> environment;
+
+    public final int STD_IN = 0;
+    public final int STD_OUT = 1;
+    public final int STD_ERR = 2;
+    private final int MAX_BUF_SIZE = 80;
+
+    // File descriptor -> buffer.
+    private final HashMap<Integer, FileBuffer> fds;
+
+    /**
+     * Flushes the buffer of a file descriptor out to its destination, and clears the buffer associated with it.
+     * For STD_IN, this just clears it.
+     * For STD_OUT and STD_ERR, this is to the screen.
+     * For any other file descriptor, this writes to that file.
+     * @param fd File descriptor to write to.
+     */
+    public void fdFlush(int fd) {
+        switch (fd) {
+            case STD_IN -> {
+                consume(STD_IN);  // just clears STD_IN
+            }
+            case STD_OUT -> System.out.print(consume(STD_OUT).orElse(""));
+            case STD_ERR -> System.out.print("\033[31m"+consume(STD_ERR).orElse("")+"\033[0m");
+            default -> {
+                var fbuf = fds.get(fd);
+                if (fbuf == null) {
+                    send(STD_ERR, "No such buffer associated with "+fd+"\n");
+                    fdFlush(STD_ERR);
+                    return;
+                }
+                if (!fbuf.mode.contains("w")) {
+                    send(STD_ERR, "File descriptor "+fd+" not open for writing\n");
+                    fdFlush(STD_ERR);
+                    return;
+                }
+
+                fbuf.file.setContents(fbuf.file.getContents() + consume(fd));
+            }
+        }
+    }
+
+    private JBashEnvironment() {
+        // Perform any initialization of the environment
+        environment = new HashMap<>();
+        environment.put("PS1", "$ ");
+        environment.put("PWD", "/");
+        environment.put("HOME", "/");
+        environment.put("?", "0");
+
+        fds = new HashMap<>();
+        fds.put(STD_IN, new FileBuffer("stdin", "rw"));
+        fds.put(STD_OUT, new FileBuffer("stdout", "rw"));
+        fds.put(STD_ERR, new FileBuffer("stderr", "rw"));
+    }
+
+    public static synchronized JBashEnvironment getInstance() {
+        if (instance == null) { instance = new JBashEnvironment(); }
+        return instance;
+    }
+
+    /**
+     * Sets environment variable.
+     * @param var Name of environment variable to set.
+     * @param val Name of value to assign <code>var</code>.
+     */
+    public void set(String var, String val) {
+        environment.put(var, val);
+    }
+
+    /**
+     * Returns a string value associated with <code>var</code>.
+     * @param var Environment variable to get value of.
+     * @return Value of <code>var</code>.
+     */
+    public String get(String var) {
+        return environment.get(var) == null
+                ? ""
+                : environment.get(var);
+    }
+
+    /**
+     * Appends a message to file descriptor <code>fd</code>'s buffer.
+     */
+    public void send(int fd, String msg) {
```

Comment:
maybe return a bool when early terminating if fd isn't found? honestly im reaching

Reply:
Funny enough I had that implemented but replaced it with void for some reason...though I much prefer Result types to booleans so we can actually know what happened. I might implement some kind of basic Result generic in Utils, so we can use that for some operations

Reply:
just return ints for allem!!!